### ISSUE TITLE
parallel_apply gets stuck forever on small dataframes

### ISSUE THREAD

**Issue by petrbel** [nalepae/pandarallel#80]
Parallel apply always gets stuck when running it on small dataframes (such as two rows). I have a suspicion that the problem occurs in the case the lenght of the dataframe is smaller than the `nb_workers`.

In tests, I can reproduce it by setting applying the following

1. `nb_workers=8` (I use 64-core server) and then
2. setting the sample dataframe length to 4

The modified part of the tests (`tests/test_pandarallel.py`):
```python
122 @pytest.fixture()
123 def pandarallel_init(progress_bar, use_memory_fs):
124     pandarallel.initialize(
125         progress_bar=progress_bar, use_memory_fs=use_memory_fs, nb_workers=8
126     )
127 
128 
129 def test_dataframe_apply_axis_0(pandarallel_init, func_dataframe_apply_axis_0):
130     df_size = int(4)
131     df = pd.DataFrame(
132         dict(
133             a=np.random.randint(1, 8, df_size),
134             b=np.random.rand(df_size),
135             c=np.random.randint(1, 8, df_size),
136             d=np.random.rand(df_size),
137             e=np.random.randint(1, 8, df_size),
138             f=np.random.rand(df_size),
139             g=np.random.randint(1, 8, df_size),
140             h=np.random.rand(df_size),
141         )
142     )
143     df.index = [item / 10 for item in df.index]
144 
145     res = df.apply(func_dataframe_apply_axis_0)
146     res_parallel = df.parallel_apply(func_dataframe_apply_axis_0)
147     assert res.equals(res_parallel)
``` 

Related git diff:
```
diff --git a/tests/test_pandarallel.py b/tests/test_pandarallel.py
index ee4e5f0..7c33fb0 100644
--- a/tests/test_pandarallel.py
+++ b/tests/test_pandarallel.py
@@ -122,12 +122,12 @@ def func_dataframe_groupby_rolling_apply(request):
 @pytest.fixture()
 def pandarallel_init(progress_bar, use_memory_fs):
     pandarallel.initialize(
-        progress_bar=progress_bar, use_memory_fs=use_memory_fs, nb_workers=2
+        progress_bar=progress_bar, use_memory_fs=use_memory_fs, nb_workers=8
     )
 
 
 def test_dataframe_apply_axis_0(pandarallel_init, func_dataframe_apply_axis_0):
-    df_size = int(1e1)
+    df_size = int(4)
     df = pd.DataFrame(
         dict(
             a=np.random.randint(1, 8, df_size),
```

Running of `pytest` then freezes (90+ minutes) like this:
```
$ pytest
============================================================================================================================= test session starts =============================================================================================================================
platform linux -- Python 3.7.2, pytest-5.1.2, py-1.8.0, pluggy-0.12.0
rootdir: /tmp/pandarallel
collected 81 items                                                                                                                                                                                                                                                            

tests/test_inliner.py .............                                                                                                                                                                                                                                     [ 16%]
tests/test_pandarallel.py ................................................

```

When I was thinking about a possible fix I came up with either:

1. When the dataframe is too short, automatically use `apply` instead of `parallel_apply`
2. Set number of pool worker to `min(nb_workers, len(df))`

To be honest, I think both of my proposed solutions handle rather the symptoms than the actual problem. I was having a lot of trouble with these edge-cases during the development of our product.

I would be happy to implement a PR once pointed to the right direction. Also please let me know if you need any further information.

Tagging my colleagues just to keep them informed: @FloopCZ @hurin @nechcigithub

**Comment by nalepae** [comment]
Fixed in **Pandarallel v1.4.7**

Thank you for you'r very well written ticket!

Please reopen if you still encounter this issue.

**Comment by petrbel** [comment]
@nalepae it works, thank you!
